lightning/src/offers/invoice_request.rs

```diff
@@ -0,0 +1,1259 @@
+		if !self.offer.supports_chain(chain) {
+			return Err(SemanticError::UnsupportedChain);
+		}
+
+		self.invoice_request.chain = Some(chain);
+		Ok(self)
+	}
+
+	#[cfg(test)]
+	pub fn chain_unchecked(mut self, network: Network) -> Self {
+		let chain = ChainHash::using_genesis_block(network);
+		self.invoice_request.chain = Some(chain);
+		self
+	}
+
+	/// Sets the [`InvoiceRequest::amount_msats`] for paying an invoice. Must be at least the base
+	/// invoice amount (i.e., [`Offer::amount`] times [`quantity`]).
+	///
+	/// Successive calls to this method will override the previous setting.
+	///
+	/// [`quantity`]: Self::quantity
+	pub fn amount_msats(mut self, amount_msats: u64) -> Result<Self, SemanticError> {
+		self.invoice_request.offer.check_amount_msats(
+			Some(amount_msats), self.invoice_request.quantity
+		)?;
+		self.invoice_request.amount_msats = Some(amount_msats);
+		Ok(self)
+	}
+
+	#[cfg(test)]
+	pub fn amount_msats_unchecked(mut self, amount_msats: u64) -> Self {
+		self.invoice_request.amount_msats = Some(amount_msats);
+		self
+	}
+
+	/// Sets the [`InvoiceRequest::features`].
+	///
+	/// Successive calls to this method will override the previous setting.
+	#[cfg(test)]
+	pub fn features(mut self, features: InvoiceRequestFeatures) -> Self {
+		self.invoice_request.features = features;
+		self
+	}
+
+	/// Sets [`InvoiceRequest::quantity`] of items. If not set, `1` is assumed. Must conform to
+	/// [`Offer::is_valid_quantity`].
+	///
+	/// Successive calls to this method will override the previous setting.
+	pub fn quantity(mut self, quantity: u64) -> Result<Self, SemanticError> {
+		self.invoice_request.offer.check_quantity(Some(quantity))?;
+		self.invoice_request.quantity = Some(quantity);
+		Ok(self)
+	}
+
+	#[cfg(test)]
+	pub fn quantity_unchecked(mut self, quantity: u64) -> Self {
+		self.invoice_request.quantity = Some(quantity);
+		self
+	}
+
+	/// Sets the [`InvoiceRequest::payer_note`].
+	///
+	/// Successive calls to this method will override the previous setting.
+	pub fn payer_note(mut self, payer_note: String) -> Self {
+		self.invoice_request.payer_note = Some(payer_note);
+		self
+	}
+
+	/// Builds an unsigned [`InvoiceRequest`] after checking for valid semantics. It can be signed
+	/// by [`UnsignedInvoiceRequest::sign`].
+	pub fn build(mut self) -> Result<UnsignedInvoiceRequest<'a>, SemanticError> {
+		#[cfg(feature = "std")] {
+			if self.offer.is_expired() {
+				return Err(SemanticError::AlreadyExpired);
+			}
+		}
+
+		let chain = self.invoice_request.chain();
+		if !self.offer.supports_chain(chain) {
+			return Err(SemanticError::UnsupportedChain);
+		}
+
+		if chain == self.offer.implied_chain() {
+			self.invoice_request.chain = None;
+		}
+
+		if self.offer.amount().is_none() && self.invoice_request.amount_msats.is_none() {
+			return Err(SemanticError::MissingAmount);
+		}
+
+		self.invoice_request.offer.check_quantity(self.invoice_request.quantity)?;
```

Comment:
I don't feel strongly, but (at least some of) these checks and maybe below seem redundant to me now(?)